nodl_python/nodl/types.py

```diff
@@ -10,13 +10,43 @@
 # You should have received a copy of the GNU Limited General Public License along with
 # this program. If not, see <http://www.gnu.org/licenses/>.
 
+from typing import Any, Dict, List
 
-from typing import List, Optional
-
+from nodl._util import qos_to_dict
 from rclpy.qos import QoSPresetProfiles, QoSProfile
 
 
-class Action:
+class NoDLData:
+    """Data structure base class for NoDL objects."""
+
+    def __repr__(self) -> str:
+        return str(self._as_dict)
+
+    def __str__(self) -> str:
+        return str(self._as_dict)
+
+    @property
+    def _as_dict(self) -> Dict:
+        self_dict = self.__dict__.copy()
+        if 'qos' in self_dict:
+            self_dict['qos'] = qos_to_dict(self_dict['qos'])
+        return self_dict
```

Comment:
I suggest leaving this as a function, not a property. This isn't necessarily a light operation, hiding it behind a property seems misleading.

Reply:
Dict.copy() is a shallow copy, its fairly cheap. Not sure its necessarily misleading.

Reply:
Alright. I don't really agree but it's not public API, I can live with it.